src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -789,6 +791,46 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
             } else {
                 debug!("intra-doc link to {} resolved to {:?}", path_str, res);
 
+                // Disallow e.g. linking to enums with `struct@`
+                if let Res::Def(kind, id) = res {
+                    debug!("saw kind {:?} with disambiguator {:?}", kind, disambiguator);
+                    // NOTE: this relies on the fact that `''` is never parsed as a disambiguator
+                    // NOTE: this needs to be kept in sync with the disambiguator parsing
+                    match (kind, disambiguator.unwrap_or_default().trim_end_matches("@")) {
+                        | (DefKind::Struct, "struct")
+                        | (DefKind::Enum, "enum")
+                        | (DefKind::Trait, "trait")
+                        | (DefKind::Union, "union")
+                        | (DefKind::Mod, "mod" | "module")
+                        | (DefKind::Const | DefKind::ConstParam | DefKind::AssocConst | DefKind::AnonConst, "const")
+                        | (DefKind::Static, "static")
+                        // NOTE: this allows 'method' to mean both normal functions and associated functions
+                        // This can't cause ambiguity because both are in the same namespace.
+                        | (DefKind::Fn | DefKind::AssocFn, "fn" | "function" | "method")
+                        | (DefKind::Macro(MacroKind::Bang), "bang")
+                        | (DefKind::Macro(MacroKind::Derive), "derive")
+                        // These are namespaces; allow anything in the namespace to match
+                        | (_, "type" | "macro" | "value")
+                        // If no disambiguator given, allow anything
+                        | (_, "")
+                        // All of these are valid, so do nothing
+                        => {}
+                        (_, disambiguator) => {
+                            // The resolved item did not match the disambiguator; give a better error than 'not found'
+                            let msg = format!("unresolved link to `{}`", path_str);
+                            report_diagnostic(cx, &msg, &item, &dox, link_range, |diag, sp| {
+                                let msg = format!("this item resolved to {} {}, which did not match the disambiguator '{}'", kind.article(), kind.descr(id), disambiguator);
```

Comment:
Do we use `this item` in any other context? IIRC most of the errors does not use `this`, maybe it would be better to just use the original item instead `{link} resolved to {} {}`? Not sure if link is the right variable.

Reply:
I would use "this link" here

Reply:
Or "this path" but I prefer "this link"